core/src/main/java/org/axonframework/common/lock/PessimisticLockFactory.java

```diff
@@ -29,46 +31,55 @@
- * thread has guaranteed unique access.
+ * {@link #obtainLock} will block until a lock could be obtained or back off based on the {@link BackoffParameters}
+ * by throwing an exception. The later will cause the command to fail, but will allow the Thread to be released. If a
+ * lock is obtained by a thread, that thread has guaranteed unique access.
  * <p/>
  * Each thread can hold the same lock multiple times. The lock will only be released for other threads when the lock
  * has been released as many times as it was obtained.
  * <p/>
  * This lock can be used to ensure thread safe access to a number of objects, such as Aggregates and Sagas.
  *
+ * Back off properties with respect to acquiring locks can be configured though the {@link BackoffParameters}.
+ *
  * @author Allard Buijze
  * @since 1.3
  */
 public class PessimisticLockFactory implements LockFactory {
 
-    private static final Set<PessimisticLockFactory> INSTANCES =
-            newSetFromMap(synchronizedMap(new WeakHashMap<>()));
+    private static final Set<PessimisticLockFactory> INSTANCES = newSetFromMap(synchronizedMap(new WeakHashMap<>()));
 
-    private static Set<Thread> threadsWaitingForMyLocks(Thread owner) {
-        return threadsWaitingForMyLocks(owner, INSTANCES);
-    }
+    private final ConcurrentHashMap<String, DisposableLock> locks = new ConcurrentHashMap<>();
+    private final BackoffParameters backoffParameters;
 
-    private static Set<Thread> threadsWaitingForMyLocks(Thread owner, Set<PessimisticLockFactory> locksInUse) {
-        Set<Thread> waitingThreads = new HashSet<>();
-        for (PessimisticLockFactory lock : locksInUse) {
-            lock.locks.values().stream()
-                    .filter(disposableLock -> disposableLock.isHeldBy(owner))
-                    .forEach(disposableLock -> disposableLock.queuedThreads().stream()
-                            .filter(waitingThreads::add)
-                            .forEach(thread -> waitingThreads.addAll(threadsWaitingForMyLocks(thread, locksInUse))));
-        }
-        return waitingThreads;
+    /**
+     * Creates a new IdentifierBasedLock instance.
+     * <p/>
+     * Deadlocks are detected across instances of the IdentifierBasedLock.
+     * This constructor specifies no back off from lock acquisition
+     *
+     * @apiNote Since the previous versions didn't support any backoff properties, this no-arg constructor creates a
+     * {@link PessimisticLockFactory} with no backoff properties. This is however a poor default (the system will
+     * very likely converge to a state where it no longer handles any commands any lock is held indefinitely.) In the
+     * next major version of Axon sane defaults should be chosen and thus behavior will change. Should your setup rely
+     * on the no-backoff behavior then you are advised to call {@link #PessimisticLockFactory(BackoffParameters)} with
+     * explicitly specified {@link BackoffParameters}.
+     */
```

Comment:
I think it is fair to add `@Deprecated` to this constructor, as that's a fair signal for us to adjust this in feature major releases.